casanovo/data/db_utils.py

```diff
@@ -0,0 +1,317 @@
+            ) * (1 - (self.precursor_tolerance / 1e6))
+
+            window = self.digest[
+                (self.digest["calc_mass"] >= lower_bound)
+                & (self.digest["calc_mass"] <= upper_bound)
+            ]
+            candidates.append(window[["peptide", "calc_mass", "protein"]])
+
+        candidates = pd.concat(candidates)
+        candidates.drop_duplicates(inplace=True)
+        candidates.sort_values(by=["calc_mass", "peptide"], inplace=True)
+        return list(candidates["peptide"]), list(candidates["protein"])
+
+    def _digest_fasta(
+        self,
+        fasta_filename: str,
+        enzyme: str,
+        digestion: str,
+        missed_cleavages: int,
+        max_mods: int,
+        min_peptide_length: int,
+        max_peptide_length: int,
+    ):
+        """
+        Digests a FASTA file and returns the peptides, their masses, and associated protein.
+
+        Parameters
+        ----------
+        fasta_filename : str
+            Path to the FASTA file.
+        enzyme : str
+            The enzyme to use for digestion.
+            See pyteomics.parser.expasy_rules for valid enzymes.
+        digestion : str
+            The type of digestion to perform. Either 'full' or 'partial'.
+        missed_cleavages : int
+            The number of missed cleavages to allow.
+        max_mods : int
+            The maximum number of modifications to allow per peptide.
+        min_peptide_length : int
+            The minimum length of peptides to consider.
+        max_peptide_length : int
+            The maximum length of peptides to consider.
+
+        Returns
+        -------
+        mod_peptide_list : pd.DataFrame
+            A Pandas DataFrame with peptide, mass,
+            and protein columns. Sorted by neutral mass in ascending order.
+        """
+        # Verify the existence of the file:
+        if not os.path.isfile(fasta_filename):
+            logger.error("File %s does not exist.", fasta_filename)
+            raise FileNotFoundError(f"File {fasta_filename} does not exist.")
+
+        fasta_data = fasta.read(fasta_filename)
+        peptide_list = []
+        if digestion not in ["full", "partial"]:
+            logger.error("Digestion type %s not recognized.", digestion)
+            raise ValueError(f"Digestion type {digestion} not recognized.")
+        semi = digestion == "partial"
+        for header, seq in fasta_data:
+            pep_set = parser.cleave(
+                seq,
+                rule=parser.expasy_rules[enzyme],
+                missed_cleavages=missed_cleavages,
+                semi=semi,
+            )
+            protein = header.split()[0]
+            for pep in pep_set:
+                if (
+                    len(pep) < min_peptide_length
+                    or len(pep) > max_peptide_length
+                ):
+                    continue
+                if any(
+                    aa in pep for aa in "BJOUXZ"
```

Comment:
Unsure what this means. This check is for unknown Amino Acids. All the others exist in the residues. Do you mean instead of checking for unknown AA's, verify that each AA is from the vocabulary?

Reply:
Yes.

Reply:
Checks for valid AAs based on the config